client/app/components/ParameterValueInput.jsx

```diff
@@ -132,25 +159,64 @@ export class ParameterValueInput extends React.Component {
   }
 
   renderNumberInput() {
-    const { value, onSelect, className } = this.props;
+    const { className, onSelect, applyButton } = this.props;
+    const { value } = this.state;
+    const showApplyButton = applyButton && value !== this.props.value;
+
+    const onChange = (newValue) => {
+      this.setState({ value: newValue });
+      if (!applyButton) {
+        onSelect(newValue);
+      }
+    };
+
     return (
-      <InputNumber
-        className={'form-control ' + className}
-        defaultValue={!isNaN(value) && value || 0}
-        onChange={onSelect}
-      />
+      <div
+        className={classNames('parameter-input-number', { 'parameter-input-number--apply-button': showApplyButton })}
```

Comment:
When it comes to element state, I tend to favor describing the state rather than it's impact (cause that usually changes as app progresses) so instead of classname `parameter-input-number--apply-button` I'd go with `input-dirty`. And if it's a bool, I'd even ditch classnames and do a `data-dirty="true"` attribute on the element.

For your consideration 👨‍💻

Reply:
Oh, this is just to have the apply button on/off 🤦‍♂️

Reply:
Hmm so I would have these as two separate attributes actually.

Where does `$ctrl.applyButton` determined? When is it `false`?

Reply:
Found it - hardcoded to `true`. Then perhaps this prop can be removed and code simplified?

Reply:
> When is it `false`?

`ParameterMappingInput` :slightly_smiling_face: (e.g: in the Add Widget Dialog). Also I thought it could be interesting to support it without the Apply Button.

Regarding the css, yes, it's to turn on/off the appearing of the apply button on the input. I tried to follow [BEM naming convention](http://getbem.com/naming/) on that, but yes, I can use a data attribute instead. It will actually look better than this `classNames` syntax.

Reply:
BTW, it was Cypress that made me realize that the Apply Button would affect the Add Widget Dialog :grin:

Reply:
> I tried to follow BEM naming convention on that

Maybe we should have a conversation on whether we want to keep using BEM.

Reply:
Updated :+1:

> Maybe we should have a conversation on whether we want to keep using BEM.

This was actually the only case where I was applying BEM (modifiers like `--apply-button`). But I preferred as @ranbena suggested `data-dirty`. I think BEM is sort of deprecated for Redash already, just need to update the docs if that's really the case.